WordPress/gutenberg
# Add defaultFontSizes theme.json (theme.json v3)

## Dev Note 📝 

See https://github.com/WordPress/gutenberg/pull/58409#issuecomment-2078077477

## What?


Reimplementation of #56661 with a proper theme.json migration.

Adds a `defaultFontSizes` option to disable showing the default font sizes.

When `defaultFontSizes === false` the options are removed from the UI.

In theme.json v3, by default `defaultFontSizes` is `true`.

In theme.json v2, `defaultFontSizes` isn't a valid key, but is treated as `false` where it would be used when a theme has font sizes defined for backwards compatibility of generated CSS.

## Why?


Having an option for `defaultFontSizes` would give themes more control over which options are available. This option brings consistency to font sizes to match colors, gradients, duotone filters, and shadow presets.

## How?


Goals are to migrate the behavior of the font sizes presets to be like other presets that can be disabled via `default*` settings.

## Testing Instructions





Screenshots below contain example modified `fontSizes` JSON for TT4 to cover various cases of having font sizes and overriding default ones.

For each modified theme.json test that each step works as expected (see screenshots):

1. Test with theme.json `"version": 2`. (trunk and this PR should be the same.)
2. Update your theme.json to `"version": 3`. (Don't forget to update the style variations too.)
3. Update your theme.json (again) with `settings.typography.defaultFontSizes` to `false`.

Then test custom origin styles migration.

1. Start on trunk with a `"version": 2` theme.json with no associated `wp_global_styles` `post_type`s in the DB.
2. Make changes to the global styles UI such as adding a new color.
3. See the `"version": 2` entry in the DB.
4. Switch to this branch.
5. Make more changes to the global styles UI such removing the color.
6. See that the entry is updated to `"version": 3` in the DB.
7. See that `"defaultFontSizes": false` is _not_ added to the entry in the DB.
8. Update the theme's theme.json to `"version": 3`.
9. Make more updates in the UI.
10. See that `"defaultFontSizes": false` is still not added.

### Testing Instructions for Keyboard


N/A

## Screenshots or screencast 

Each set of screenshots shows a different example theme.json modification to TT4.

- v2 trunk: Screenshot of previous behavior in v2 theme.json.
- v2: Screenshot from this branch for v2 theme.json.
- v3 (default): Screenshot with v3 theme.json. (Should be the same as `"defaultFontSizes": true`.)
- v3 `false`: Screenshot with `"defaultFontSizes": false` set in v3 theme.json.

### TT4 Footnote Override

<details>
<summary>Example JSON</summary>

```jsonc
// Override fontSizes
// Same as TT4 with small renamed to footnote
{
	"settings": {
		"typography": {
			"fontSizes": [
				{
					"fluid": false,
					"name": "Footnote",
					"size": "0.9rem",
					"slug": "footnote"
				},
				{
					"fluid": false,
					"name": "Medium",
					"size": "1.05rem",
					"slug": "medium"
				},
				{
					"fluid": {
						"min": "1.39rem",
						"max": "1.85rem"
					},
					"name": "Large",
					"size": "1.85rem",
					"slug": "large"
				},
				{
					"fluid": {
						"min": "1.85rem",
						"max": "2.5rem"
					},
					"name": "Extra Large",
					"size": "2.5rem",
					"slug": "x-large"
				},
				{
					"fluid": {
						"min": "2.5rem",
						"max": "3.27rem"
					},
					"name": "Extra Extra Large",
					"size": "3.27rem",
					"slug": "xx-large"
				}
			]
		}
	}
}
```

</details>

| v2 trunk | v2 | v3 (default) | v3 `false` |
| -------- | -- | ------------ | ---------- |
| ![font-sizes-v2-override-before](https://github.com/WordPress/gutenberg/assets/5129775/a0360067-87f1-4b0d-b489-275eec6fb969) | ![font-sizes-v2-override-after](https://github.com/WordPress/gutenberg/assets/5129775/7a52e2cd-dbd7-4e13-a284-43f2fa351433) | ![font-sizes-v3-override-true](https://github.com/WordPress/gutenberg/assets/5129775/af80a3fe-7926-4a10-85e5-85bca5406eb9) | ![font-sizes-v3-override-false](https://github.com/WordPress/gutenberg/assets/5129775/46dfb059-6fdf-4560-b3b5-fd645094884d) |

### TT4 Custom Only

<details>
<summary>Example JSON</summary>

```jsonc
// Only custom fontSizes
// Same as override with only footnote and extra-extra-large
{
	"settings": {
		"typography": {
			"fontSizes": [
				{
					"fluid": false,
					"name": "Footnote",
					"size": "0.9rem",
					"slug": "footnote"
				},
				{
					"fluid": {
						"min": "2.5rem",
						"max": "3.27rem"
					},
					"name": "Extra Extra Large",
					"size": "3.27rem",
					"slug": "xx-large"
				}
			]
		}
	}
}
```

</details>

| v2 trunk | v2 | v3 (default) | v3 `false` |
| -------- | -- | ------------ | ---------- |
| ![font-sizes-v2-custom-before](https://github.com/WordPress/gutenberg/assets/5129775/9310f4c1-37c3-42d4-bef8-efe7c7b08e99) | ![font-sizes-v2-custom-after](https://github.com/WordPress/gutenberg/assets/5129775/e8039677-055f-4c9b-8731-4f3480721c25) | ![font-sizes-v3-custom-true](https://github.com/WordPress/gutenberg/assets/5129775/eb3c6c75-cdb9-42e6-b03d-21abee3eb2e3) | ![font-sizes-v3-custom-false](https://github.com/WordPress/gutenberg/assets/5129775/435fb7ff-0b11-4dbc-8b56-8b1f244b9d25) |

### TT4 Empty `fontSizes`

<details>
<summary>Example JSON</summary>

```jsonc
// Empty fontSizes
{
	"settings": {
		"typography": {
			"fontSizes": []
		}
	}
}
```

</details>

| v2 trunk | v2 | v3 (default) | v3 `false` |
| -------- | -- | ------------ | ---------- |
| ![font-sizes-v2-empty-before](https://github.com/WordPress/gutenberg/assets/5129775/bc992dd4-d711-4942-ad32-dcd570ffd2e6) | ![font-sizes-v2-empty-after](https://github.com/WordPress/gutenberg/assets/5129775/6d776bf3-13ab-461f-955e-f6e35f246c8b) | ![font-sizes-v3-empty-true](https://github.com/WordPress/gutenberg/assets/5129775/565c07b7-9c5f-47fa-9395-2532617a32db) | ![font-sizes-v3-empty-false](https://github.com/WordPress/gutenberg/assets/5129775/e613f04f-bf8b-4a93-bf45-686b2deaa80f) |

### TT4 Undefined `fontSizes`

<details>
<summary>Example JSON</summary>

```jsonc
// Undefined fontSizes
{
	"settings": {
		"typography": {}
	}
}
```

</details>

| v2 trunk | v2 | v3 (default) | v3 `false` |
| -------- | -- | ------------ | ---------- |
| ![font-sizes-v2-undefined-before](https://github.com/WordPress/gutenberg/assets/5129775/461deb78-a6e2-4c20-9417-1243f4ed0f44) | ![font-sizes-v2-undefined-after](https://github.com/WordPress/gutenberg/assets/5129775/a300104d-6d22-4ff9-9e91-9f7a69b5059f) | ![font-sizes-v3-undefined-true](https://github.com/WordPress/gutenberg/assets/5129775/2ebde0c6-26d9-44bb-bff1-c11181eafcef) | ![font-sizes-v3-undefined-false](https://github.com/WordPress/gutenberg/assets/5129775/4aa5d285-d605-4edd-9516-2358dc8873cd) |
